### PR TITLE
use new juju/http pkg in juju

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Start using juju/http  for new http clients, SecureTLSConfig and BasicAuthHeader.  This functionality has been moved from the juju/utils package and enhanced.  

- HTTP clients are now constructed with config, rather than many constructor methods.  
- Clients are created with current proxy settings, not those cached when the jujud started, resolving 1798734.
- We can getting logging from the client, currently only used by simplestreams.
- Get() takes a context for future improvements.
- Get() will also use the httptrace package to display details on the connection, if log level is set to Trace in config.  

To enable the tracing here, add "juju.environs.simplestreams.http=TRACE" to the logging-config during bootstrap or at model-config

Additional changed in the future should be made to use httptest more instead of the over ride of the http DefaultTransport and RoundTripper for test scenarios.

## QA steps

```console
# bootstrap with logging enabled
$ juju bootstrap <openstack cloud> test  --logging-config="<root>=DEBUG;juju.environs.simplestreams.http=TRACE" --show-log --config logging-config="<root>=DEBUG;juju.environs.simplestreams.http=TRACE"
...
16:01:48 TRACE juju.environs.simplestreams.http client.go:174 request for "http://cloud-images.ubuntu.com/releases/streams/v1/index2.sjson": "GET /releases/streams/v1/index2.sjson HTTP/1.1\r\nHost: cloud-images.ubuntu.com\r\nUser-Agent: Go-http-client/1.1\r\nAccept-Encoding: gzip\r\n\r\n"
...

# change a juju proxy setting and see things not work
$ juju model-config  -m controller  juju-http-proxy="http://squid.internal:3128/"
$ juju deploy ubuntu
....
ERROR cannot add application "ubuntu"
....

# set back and see juju deploy
$ juju model-config  -m controller  juju-http-proxy=""
$ juju deploy ubuntu
Located charm "cs:ubuntu-15".
Deploying charm "cs:ubuntu-15".

# check logs
$ juju debug-log -m controller --include-module  juju.environs.simplestreams.http --replay 
... 
16:19:54 TRACE juju.environs.simplestreams.http client.go:177 request for "https://streams.canonical.com/juju/tools/streams/v1/index2.sjson": "GET /juju/tools/streams/v1/index2.sjson HTTP/1.1\r\nHost: streams.canonical.com\r\nUser-Agent: Go-http-client/1.1\r\nAccept-Encoding: gzip\r\n\r\n"
16:19:54 TRACE juju.environs.simplestreams.http client.go:189 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson Get Conn: "streams.canonical.com:443"
16:19:54 TRACE juju.environs.simplestreams.http client.go:180 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson DNS Start: "streams.canonical.com"
16:19:54 TRACE juju.environs.simplestreams.http client.go:183 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson DNS Info: {Addrs:[{IP:2001:67c:1360:8001:1::5 Zone:} {IP:91.189.88.141 Zone:}] Err:<nil> Coalesced:false}
16:19:54 TRACE juju.environs.simplestreams.http client.go:186 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson Connection Done: network "tcp", addr "[2001:67c:1360:8001:1::5]:443", err %!q(<nil>)
16:19:54 TRACE juju.environs.simplestreams.http client.go:195 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson TLS Handshake Start
16:19:54 TRACE juju.environs.simplestreams.http client.go:198 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson TLS Handshake Done: complete true, verified chains 1, server name ""
16:19:54 TRACE juju.environs.simplestreams.http client.go:192 https://streams.canonical.com/juju/tools/streams/v1/index2.sjson Got Conn: {Conn:0xc00020f180 Reused:false WasIdle:false IdleTime:0s}
16:19:54 TRACE juju.environs.simplestreams.http client.go:177 request for "https://streams.canonical.com/juju/tools/streams/v1/mirrors.sjson": "GET /juju/tools/streams/v1/mirrors.sjson HTTP/1.1\r\nHost: streams.canonical.com\r\nUser-Agent: Go-http-client/1.1\r\nAccept-Encoding: gzip\r\n\r\n"
16:19:54 TRACE juju.environs.simplestreams.http client.go:189 https://streams.canonical.com/juju/tools/streams/v1/mirrors.sjson Get Conn: "streams.canonical.com:443"
16:19:54 TRACE juju.environs.simplestreams.http client.go:192 https://streams.canonical.com/juju/tools/streams/v1/mirrors.sjson Got Conn: {Conn:0xc00020f180 Reused:true WasIdle:true IdleTime:2.113078ms}

```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1798734
Also found during investigation of https://bugs.launchpad.net/juju/+bug/1885145, causing delays in resolution.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -154,3 +154,5 @@ replace gopkg.in/yaml.v2 => github.com/juju/yaml v0.0.0-20200420012109-12a32b78d
 replace github.com/dustin/go-humanize v1.0.0 => github.com/dustin/go-humanize v0.0.0-20141228071148-145fabdb1ab7
 
 replace github.com/hashicorp/raft-boltdb => github.com/juju/raft-boltdb v0.0.0-20200518034108-40b112c917c5
+
+replace github.com/juju/http => /Users/heathermlanigan/gohack/github.com/juju/http
```
Comment (SimonRichardson) [inline go.mod:158]: Fix this once the other PR lands.

**Comment by hmlanigan** [comment]
rebased develop and updated to juju/http, pr merged.

**Comment by hmlanigan** [comment]
Needs https://github.com/juju/http/pull/2 to land so go.mod can be updated again.  Issue found in QA once the steps were updated.

File: go.mod (hunk 1 above)
Comment (hmlanigan) [reply go.mod]: done

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**